### PR TITLE
Added ability to create textures from HTML5 canvases

### PR DESCRIPTION
**Pull request by MatthewShotton**
**Description:**
Currently it seems only images and videos can be used as texture sources for entities.

This PR adds the ability to use a canvas as the texture source for an entity. It includes a less-than-graceful hack so that THREE.js treats the canvas as a video and handles the continuous update of it.

**Changes proposed:**
- adds loadCanvas to systems/material.js
- adds isCanvas to validateSrc in /utils/src-loader.js
- adds an example.

Kind regards,
Matt

### CODE REVIEW

File: src/utils/src-loader.js [hunk 1]
```diff
@@ -37,11 +39,13 @@ function validateSrc (src, isImageCb, isVideoCb) {
...
   isImage = textureEl && textureEl.tagName === 'IMG';
   isVideo = textureEl && textureEl.tagName === 'VIDEO';
+  isCanvas = textureEl && textureEl.tagName === 'CANVAS';
   if (isImage) { return isImageCb(textureEl); }
   if (isVideo) { return isVideoCb(textureEl); }
+  if (isCanvas) { return isCanvasCb(textureEl); }
 
-  // src is a valid selector but doesn't match with a <img> or <video> element.
-  warn('"%s" does not point to a valid <img> or <video> element', src);
+  // src is a valid selector but doesn't match with a <img>, <video>, or <cavnas> element.
```
Comment (dmarcos) [inline src/utils/src-loader.js]: typo `cavnas`

**Comment by ngokevin** [comment]
Pretty awesome! People have been registering custom materials, but this seems like a much simpler API.

File: src/systems/material.js [hunk 2]
```diff
@@ -63,6 +64,21 @@ module.exports.System = registerSystem('material', {
...
 
+  /**
+   * High-level function for loading canvas textures (THREE.Texture).
+   *
+   * @param {Element|string} src - Texture source.
+   * @param {object} data - Texture data.
+   * @param {function} cb - Callback to pass texture to.
+   */
+  loadCanvas: function (src, data, cb) {
+    // duck-punch a readyState and a HAVE_CURRENT_DATA attiribute onto the canvas element so THREE.videoTexture will work
```
Comment (ngokevin) [inline src/systems/material.js]: `Hack readyState and HAVE_CURRENT_DATA on canvas to work with THREE.VideoTexture`

File: src/systems/material.js [hunk 3]
```diff
@@ -63,6 +64,21 @@ module.exports.System = registerSystem('material', {
...
+    src.readyState = 2;
+    src.HAVE_CURRENT_DATA = 2;
+    this.loadVideo(src, data, cb);
+    return;
```
Comment (ngokevin) [inline src/systems/material.js]: unecessary return

File: examples/test/canvas/index.html [hunk 4]
```diff
@@ -0,0 +1,51 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <title>Video</title>
+    <meta name="description" content="Video - A-Frame">
```
Comment (ngokevin) [inline examples/test/canvas/index.html]: Title `Canvas Texture` and folder `canvas-texture`

File: examples/test/canvas/index.html [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    <script src="../../../dist/aframe.js"></script>
+  </head>
+  <body>
+    <script type="text/javascript">
+      window.onload = function(){
```
Comment (ngokevin) [inline examples/test/canvas/index.html]: Extra credit to use the component system to hook into A-Frame's requestAnimationFrame and for reusability.

```js
AFRAME.registerComponent('update-canvas', {
  schema: {default: ''},

  init: function () {
    this.canvas = document.getElementById(this.data);
    this.ctx = canvas.getContext('2d');
  },

  tick: function (t) {
    var hue = t / 10;
    // ...
  }
});
```

```html
<a-scene update-canvas="helloworld-canvas">
</a-scene>
```

**Comment by ngokevin** [comment]
Should be good after changes, thanks!

**Comment by MatthewShotton** [comment]
Hey up, I've made those changes. Would you prefer I collapse all changes to a single commit and force-push or is multiple commits OK? (BTW, cheers for the ace library!)

**Comment by ngokevin** [comment]
Thanks! I can squash-merge.

**Pull request merged**